### DISCUSSION TITLE
`Assert.Pass` API

### DISCUSSION THREAD

**Discussion by WeihanLi** [xunit/xunit#2855] [Question] [answered]
Since we had introduced `Assert.Fail()`, could we consider introduce `Assert.Pass()` also

When the test method does not have an `Assert`, the sonar analyzer would treat it as a code smell

see: 
https://sonarcloud.io/organizations/iherbllc/rules?open=csharpsquid%3AS2699&rule_key=csharpsquid%3AS2699

I want to add the `Assert.Pass()` for verifing the test method executed without any exception

It relates to https://github.com/xunit/xunit/discussions/2637

**Comment by bartelink** [comment]
That would sort of miss the point of the alert.
As much as possible, xUnit tries to give you tools to do AAA testing. Having an `Assert.Cheating()` method is simply not part of that.

IME one of the common cases where people want this operator, is where there's an exception, and to signal that this was expected, they put a Fail in the try side, and a Pass in the catch. In that instance, the better pattern is to use Record.Exception to trap it, and then Assert against that, preserviing AAA symmetry.

Perhaps you can share what you believe to be 2 or 3 use cases where the only possible answer is to have an Assert.CheatMode operator of this kind (and hopefully people can suggest refactorings to avoid doing that)

**Reply by WeihanLi** [reply]
Thanks for the detailed reply.

For example, sometimes we may want to add a test case for the `Dispose` method

```c#
public class TestClass : IDisposable
{
    private Connection? _connection;

    public void Initialize()
    {
        _connection = new Connection();
    }

    public void Dispose()
    {
        _connection?.Dispose();
    }

}
```

When my test case may likes below:

```c#
[Fact]
public void Dispose_NoConnection_Test()
{
    var t = new TestClass();
    t.Dispose();
}
```

**Reply by bartelink** [reply]
Hm, there used to be a DoesNoThrow, but it was removed: https://github.com/xunit/xunit/issues/188

Regarding testing things like this don't break, not sure I've seen a good treatment of the tradeoffs and/or how to think about it but I wold tend not to explicitly test it - I'd use `using` the thing, and trust that the various cases (testing double disposal etc) will cover that sufficiently.

Just calling a method without any assertions and/or any way to prove it achieved its desired effect is a rare thing, and it's easy to fall into lazy thinking about how one might verify something works (or whether a given scenario is worthy of testing.)

**Reply by bradwilson** [reply, marked as answer]
The equivalent is to wrap the line you expect not to throw, and make sure it didn't throw.

```csharp
[Fact]
public void Dispose_NoConnection_Test()
{
    var t = new TestClass();

    var ex = Record.Exception(() => t.Dispose());

    Assert.Null(ex);
}
```

**Reply by WeihanLi** [reply]
Thanks, it works for me, think `Assert.Pass()` may make it more simple since we do not have to `Record.Exception`

for my case, it may look like below:

```c#
public class TestClass : IAsyncDisposable
{
    private Connection? _connection;

    public void Initialize()
    {
        _connection = new Connection();
    }

    public async ValueTask DisposeAsync()
    {
        if (_connection != null)
        {
            await _connection.DisposeAsync();
        }
    }
}

[Fact]
public async Task DisposeTest_NoConnection()
{
    var test = new TestClass();
    var exception = await Record.ExceptionAsync(async () => await test.DisposeAsync());
    Assert.Null(exception);
}
```

With `Assert.Pass`

```c#
[Fact]
public async Task DisposeTest_NoConnection()
{
    var test = new TestClass();
    await test.DisposeAsync();
    Assert.Pass();
}
```

**Reply by bradwilson** [reply]
`Assert.Pass()` would literally do nothing, so we won't be adding it.

If you'd like to extend `Assert` to add an `Assert.Pass()` for your own projects (by creating a new assertion library based on the source from ours), we have a [sample project](https://github.com/xunit/samples.xunit/tree/main/AssertSource) that shows how you can do that.
